File: Data Structures/Linked Lists/Java/DoublyLinkedList.java
import java.util.EmptyStackException;

public class DoublyLinkedList<T> {
    private Node head, tail;
    private int size;

    public DoublyLinkedList() {
    }

    public DoublyLinkedList(T data) {
        this.head = new Node(data);
        tail = head;
    }

    public void append(T data) {
        if (head == null) {
            head = tail = new Node(data);
        } else {
            tail.next = new Node(data, null, tail);
            tail = tail.next;
        }
        size++;
    }

    public void insert(int pos, T data) {
        if (pos < 0 || pos > this.size)
            throw new IndexOutOfBoundsException();
        if (pos == 0)
            insertAtHead(data);
        else if (pos == this.size)
            append(data);
        else {
            Node curr = head;
            for (int i = 0; i < pos - 1; i++)
                curr = curr.next;
            Node newNode = new Node(data, curr.next, curr);
            curr.next.prev = newNode;
            curr.next = newNode;
            size++;
            return;
        }
    }

    private void insertAtHead(T data) {
        Node newNode = new Node(data, head, null);
        newNode.prev = head;
        head = newNode;
        size++;
    }

    public T pop() {
        if (isEmpty())
            throw new RuntimeException("Empty List");
        T data = tail.data;
        tail = tail.prev;
        --size;
        if (isEmpty())
            head = null;
        else
            tail.next = null;
        return data;
    }

    public T remove(int pos) {
        if (isEmpty())
            throw new RuntimeException("Empty List");
        if (pos < 0 || pos >= this.size)
            throw new IndexOutOfBoundsException();
        if (pos == this.size - 1)
            return pop();
        if (pos == 0)
            return removeHead();
        Node curr = head;
        for (int i = 0; i < pos - 1; i++)
            curr = curr.next;
        Node remove = curr.next;
        curr.next.next.prev = curr;
        curr.next = curr.next.next;
        remove.next = null;
        remove.prev = null;
        --size;
        if (isEmpty())
            head = null;
        return remove.data;
    }

    private T removeHead() {
        T data = head.data;
        head = head.next;
        head.prev = null;
        --size;
        if (isEmpty())
            head = null;
        return data;
    }

    public int size() {
        return this.size;
    }

    public boolean isEmpty() {
        return this.size == 0;
    }

    @Override
    public String toString() {
        if (head == null)
            return "[]";
        Node curr = head;
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        while (curr.next != null) {
            sb.append(curr.data).append(", ");
            curr = curr.next;
        }
        sb.append(curr.data).append("]");
        return sb.toString();
    }

    private class Node {
        T data;
        Node next, prev;

        private Node(T data) {
            this.data = data;
        }

        private Node(T data, Node next, Node prev) {
            this.data = data;
            this.next = next;
            this.prev = prev;
        }
    }
}

File: Data Structures/Linked Lists/Python/DoublyLinkedList.py
class DoublyLinkedList:
    def __init__(self, data=None, next=None, prev=None):
        self.head = self.tail = self.Node(data, next, prev)
        self._size = 0

    def append(self, data):
        if self.isEmpty():
            self.head = self.tail = self.Node(data)
        else:
            self.tail.next = self.Node(data, None, self.tail)
            self.tail = self.tail.next
        self._size += 1

    def insert(self, index, data):
        if index < 0 or index > self._size: raise IndexError()
        elif index == self._size:
            self.append(data)
        elif index == 0:
            self._insert_at_head(data)
        else:
            curr = self.head
            i = 0
            while i != index - 1:
                curr = curr.next
                i += 1
            newNode = self.Node(data, curr.next, curr)
            curr.next.prev = newNode
            curr.next = newNode
            self._size += 1

    def _insert_at_head(self, data):
        dummy = self.head
        self.head = self.Node(data, dummy, None)
        dummy.prev = self.head
        self._size += 1

    def pop(self):
        if self.isEmpty(): raise ValueError("Cannot remove from empty list")
        node = self.tail
        self.tail = self.tail.prev
        node.next = None
        node.prev = None
        self._size -= 1
        return node.data

    def remove(self, index):
        if self.isEmpty(): raise ValueError("Cannot remove from empty list")
        elif index < 0 or index >= self._size: raise IndexError()
        elif index == 0:
            return self._remove_at_head()
        elif index == self._size - 1:
            return self.pop()
        else:
            i = 0
            curr = self.head
            while i != index - 1:
                curr = curr.next
                i += 1
            node = curr.next
            curr.next = curr.next.next
            curr.next.prev = curr
            self._size -= 1
            node.next = None
            node.prev = None
            return node.data

    def _remove_at_head(self):
        node = self.head
        self.head = self.head.next
        self.head.prev = None
        node.next = None
        node.prev = None
        self._size -= 1
        return node.data

    def size(self):
        return self._size

    def isEmpty(self):
        return self._size == 0
    
    def __str__(self):
        arr = []
        if not self.isEmpty():
            curr = self.head
            for i in range(self._size):
                arr.append(curr.data)
                curr = curr.next
        return arr.__str__()

    class Node:
        def __init__(self, data=None, next=None, prev=None):
            self.data = data
            self.next = next
            self.prev = prev